File: src/kv-store.ts
import fs from 'node:fs/promises';
import path from 'node:path';
import { Mutex } from 'async-mutex';

type KeyValueEntry = {
	key: string[];
	value: string;
};

export class KeyValueStore {
	private storePath: string | null = null;
	private entries: KeyValueEntry[] = [];

	private readonly mutex = new Mutex();

	async init(storePath: string): Promise<void> {
		return this.mutex.runExclusive(async () => {
			if (this.storePath !== null) {
				throw new Error('The KeyValueStore is already initialized');
			}
			this.storePath = storePath;

			if (await fs.exists(storePath)) {
				const entriesString = await fs.readFile(storePath, 'utf-8');
				this.entries = JSON.parse(entriesString);
			}
		});
	}

	private keyEquals(a: string[], b: string[]): boolean {
		if (a.length !== b.length) {
			return false;
		}

		return a.every((v, i) => v === b[i]);
	}

	async get(key: string[]): Promise<string | null> {
		return this.mutex.runExclusive(() => {
			if (this.storePath === null) {
				throw new Error('The KeyValueStore is not initialized yet');
			}

			const entry = this.entries.find((e) => this.keyEquals(e.key, key));
			return entry?.value ?? null;
		});
	}

	async getByPrefix(prefix: string[]): Promise<string[]> {
		return this.mutex.runExclusive(() => {
			if (this.storePath === null) {
				throw new Error('The KeyValueStore is not initialized yet');
			}

			return this.entries
				.filter((e) => this.keyEquals(e.key.slice(0, prefix.length), prefix))
				.map((e) => e.value);
		});
	}

	async set(key: string[], value: string): Promise<void> {
		return this.mutex.runExclusive(async () => {
			if (this.storePath === null) {
				throw new Error('The KeyValueStore is not initialized yet');
			}

			this.entries = this.entries.filter((e) => !this.keyEquals(e.key, key));
			this.entries.push({ key, value });

			const directory = path.dirname(this.storePath);
			const directoryExists = await fs.exists(directory);
			if (directoryExists === false) {
				await fs.mkdir(directory, { recursive: true });
			}

			const jsonString = JSON.stringify(this.entries, null, 2);
			await fs.writeFile(this.storePath, jsonString, 'utf-8');
		});
	}
}

File: kv_store.go
package main

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"sync"
)

type KeyValueEntry struct {
	Key   []string `json:"key"`
	Value string   `json:"value"`
}

type KeyValueStore struct {
	storePath   string
	entries     []KeyValueEntry
	mu          sync.Mutex
	initialized bool
}

func NewKeyValueStore(storePath string) (*KeyValueStore, error) {
	var kvs = &KeyValueStore{storePath: storePath}

	kvs.mu.Lock()
	defer kvs.mu.Unlock()

	if kvs.initialized {
		return nil, errors.New("the KeyValueStore is already initialized")
	}
	kvs.initialized = true
	kvs.storePath = storePath

	if _, err := os.Stat(storePath); err == nil {
		data, err := os.ReadFile(storePath)
		if err != nil {
			return nil, err
		}
		err = json.Unmarshal(data, &kvs.entries)
		if err != nil {
			return nil, err
		}
	}

	return kvs, nil
}

func keyEquals(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func (kvs *KeyValueStore) Get(key []string) (string, error) {
	kvs.mu.Lock()
	defer kvs.mu.Unlock()

	if !kvs.initialized {
		return "", errors.New("the KeyValueStore is not initialized yet")
	}

	for _, entry := range kvs.entries {
		if keyEquals(entry.Key, key) {
			return entry.Value, nil
		}
	}

	return "", nil
}

func (kvs *KeyValueStore) GetByPrefix(prefix []string) ([]string, error) {
	kvs.mu.Lock()
	defer kvs.mu.Unlock()

	if !kvs.initialized {
		return nil, errors.New("the KeyValueStore is not initialized yet")
	}

	var values []string
	for _, entry := range kvs.entries {
		if keyEquals(entry.Key[:len(prefix)], prefix) {
			values = append(values, entry.Value)
		}
	}
	return values, nil
}

func (kvs *KeyValueStore) Set(key []string, value *string) error {
	kvs.mu.Lock()
	defer kvs.mu.Unlock()

	if !kvs.initialized {
		return errors.New("the KeyValueStore is not initialized yet")
	}

	// Remove existing entry with the same key
	for i := 0; i < len(kvs.entries); i++ {
		if keyEquals(kvs.entries[i].Key, key) {
			kvs.entries = append(kvs.entries[:i], kvs.entries[i+1:]...)
			break
		}
	}

	if value != nil {
		kvs.entries = append(kvs.entries, KeyValueEntry{Key: key, Value: *value})
	}

	// Ensure the directory exists
	dir := filepath.Dir(kvs.storePath)
	if err := os.MkdirAll(dir, os.ModePerm); err != nil {
		return err
	}

	// Write entries to file
	data, err := json.MarshalIndent(kvs.entries, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(kvs.storePath, data, os.ModePerm)
}
